src/apps/activities/api/activities.py

```diff
@@ -358,3 +360,78 @@ async def __filter_activities(
                     activities.remove(activity)
 
     return activities
+
+
+async def applet_activities_counters_for_subject(
+    applet_id: uuid.UUID,
+    subject_id: uuid.UUID,
+    user: User = Depends(get_current_user),
+    language: str = Depends(get_language),
+    session=Depends(get_session),
+    answer_session=Depends(get_answer_session),
+) -> Response[ActivitiesCounters]:
+    applet_service = AppletService(session, user.id)
+    await applet_service.exist_by_id(applet_id)
+    await CheckAccessService(session, user.id).check_applet_detail_access(applet_id)
+
+    await SubjectsService(session, user.id).exist_by_id(subject_id)
+
+    # Fetch assigned activity or flow IDs for the subject
+    assigned_ids = await ActivityAssignmentService(session).get_assigned_activity_or_flow_ids_for_subject(subject_id)
+
+    # Fetch activities submissions by the subject
+    submitted_activities = await AnswerService(session, user.id, answer_session).get_submissions_by_subject(subject_id)
+
+    # Fetch auto assigned activity and flow IDs by applet ID
+    auto_activity_ids = await ActivityService(session, user.id).get_activity_and_flow_ids_by_applet_id_auto(applet_id)
+
+    activity_ids_as_respondent = set(assigned_ids.as_respondent + auto_activity_ids)
+    activity_ids_as_target = set(assigned_ids.as_target + auto_activity_ids)
+    all_assigned_ids = activity_ids_as_respondent.union(activity_ids_as_target)
+
+    activities_or_flows_ids_with_submissions = set(submitted_activities.activities.keys())
+    activities_or_flows_ids_without_submissions = all_assigned_ids - activities_or_flows_ids_with_submissions
+
+    activities_counters = ActivitiesCounters(
+        subject_id=subject_id,
+        respondent_activities_count=len(activity_ids_as_respondent),
+        target_activities_count=len(activity_ids_as_target),
+    )
+
+    # Add activities without submissions
+    for activity_or_flow_id in activities_or_flows_ids_without_submissions:
+        activities_counters.activities_or_flows.append(
+            ActivitySubjectCounters(
+                activity_or_flow_id=activity_or_flow_id,
+                respondents_count=0,
+                subjects_count=0,
+                respondent_submissions_count=0,
+                subject_submissions_count=0,
+            )
+        )
```

Comment:
I'm not sure I understand this logic. The variable `activities_or_flows_ids_without_submissions` represents **all _assigned_ activity/flow IDs** for which there happens to be no submissions yet. If these are activities that have been assigned, then why are we setting `respondents_count` and `subjects_count` to zero? At least one of those values has to be nonzero, else they couldn't have ended up in the `all_assigned_ids` set, which is what this set is derived from.

This endpoint should in fact never return any activity or flow for which all values are zero. If it is returning an activity or flow for which all values are zero, then that activity or flow has no relationship with the subject ID that was passed to the endpoint, which would mean some part of the logic that produces these sets of activity/flow IDs is wrong. These activities/flows **have to be related to the subject ID** you passed in, either by assignment (as respondent and/or subject), or by submission (as respondent and/or subject), or both.

Reply:
Currently, I'm getting 1 result from this endpoint on my localhost that has all values set to zero. I don't know why it's being returned yet, I haven't dug deeply to figure it out, but I'll start.

Reply:
I'm currently getting one result for one of my subjects for which all values are zero… I have to dig a little more deeply still but I'm assuming it's because there are no submissions, but there are active assignments, so I'll see what the return value is supposed to be and follow up.

Reply:
So… I think I see some of the confusion. Looking further down in the second loop below, on lines 430-431:

```tsx
                respondents_count=len(activity_data.respondents),
                subjects_count=len(activity_data.subjects),
```

We're only counting subjects and respondents coming from _submission data_. But we _also_ have to include subjects and respondents from _assignments_:
- `subjects_count` should be the total count of the _union_ of the set of subjects who have submissions as the target + **the set of subjects who are assigned as targets**
- `respondents_count` should be the total count of the _union_ of the set of subjects who have submissions as the respondent + **the set of subjects who are assigned as respondents**

Does that make sense?

Reply:
I rewrite this code